src/coreclr/vm/exceptionhandling.cpp

```diff
@@ -7679,6 +7679,28 @@ UINT_PTR GetEstablisherFrame(REGDISPLAY* pvRegDisplay, ExInfo* exInfo)
 #endif
 }
 
+size_t GetSSPForFrameOnCurrentStack(CONTEXT *pContext)
+{
+    size_t *targetSSP = 0;
+
+#if defined(HOST_AMD64) && defined(HOST_WINDOWS)
+    targetSSP = (size_t *)_rdsspq();
+    // Find the shadow stack pointer for the frame we are going to restore to.
+    // The SSP we search is pointing to the return address of the frame represented
+    // by the passed in context. So we search for the instruction pointer from 
+    // the context and return one slot up from there.
+    if (targetSSP != NULL)
+    {
+        size_t ip = GetIP(pContext);
+        while (*targetSSP++ != ip)
+        {
```

Comment:
is there always a guarantee that we are going to find the right SSP here?

Reply:
Unless the shadow stack is corrupted, it has to be there as we got that address by unwinding regular stack. If shadow stack is corrupted, we would walk out of it here and crash with access violation. But there is no way known to me to get the bounds of the shadow stack, unless we want to scan memory for a guard page, which looks like an overkill.

Reply:
ok seems reasonable.